include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+        //union {
+            STextureSource texture;
+        //};
+    };
+
+    struct CEmissionNode : INode
+    {
+        CEmissionNode() : INode(ES_EMISSION) {}
+
+        float intensity;
+    };
+
+    struct CBSDFCombinerNode : INode
+    {
+        enum E_TYPE
+        {
+            //mix of N BSDFs
+            ET_MIX,
+            //blend of 2 BSDFs weighted by constant or texture
+            ET_WEIGHT_BLEND,
+            //blend of 2 BSDFs weighted by fresnel
+            ET_FRESNEL_BLEND,
+            //blend of 2 BSDFs weighted by custom direction-based curve
+            ET_CUSTOM_CURVE_BLEND,
+            ET_DIFFUSE_AND_SPECULAR
+        };
+
+        E_TYPE type;
+
+        CBSDFCombinerNode(E_TYPE t) : INode(ES_BSDF_COMBINER), type(t) {}
+    };
+    struct CBSDFBlendNode : CBSDFCombinerNode
+    {
+        CBSDFBlendNode() : CBSDFCombinerNode(ET_WEIGHT_BLEND) {}
+
+        SParameter<float> weight;
+    };
+    struct CBSDFMixNode : CBSDFCombinerNode
+    {
+        CBSDFMixNode() : CBSDFCombinerNode(ET_MIX) {}
+
+        float weights[MAX_CHILDREN];
+    };
+
+    struct CBSDFNode : INode
+    {
+        enum E_TYPE
+        {
+            ET_DIFFTRANS,
+            ET_SPECULAR_DELTA,
+            ET_MICROFACET_DIFFUSE,
+            ET_MICROFACET_SPECULAR,
+            ET_SHEEN,
+            ET_COATING
+        };
+
+        CBSDFNode(E_TYPE t) :
+            INode(ES_BSDF),
+            type(t),
+            eta(1.f,1.f,1.f),
+            etaK(0.f,0.f,0.f)
+        {}
+
+        E_TYPE type;
+        color_t eta, etaK;
+    };
+    struct CMicrofacetSpecularBSDFNode : CBSDFNode
+    {
+        enum E_NDF
+        {
+            ENDF_BECKMANN,
+            ENDF_GGX,
+            ENDF_ASHIKHMIN_SHIRLEY,
+            ENDF_PHONG
+        };
+        enum E_SHADOWING_TERM
+        {
+            EST_SMITH,
+            EST_VCAVITIES
+        };
+        enum E_SCATTER_MODE
+        {
+            ESM_REFLECT,
+            ESM_TRANSMIT
+        };
+
+        CMicrofacetSpecularBSDFNode(E_TYPE t = ET_MICROFACET_SPECULAR) : CBSDFNode(t) {}
+
+        void setSmooth(E_NDF _ndf = ENDF_GGX)
+        {
+            ndf = _ndf;
+            alpha_u.source = EPS_CONSTANT;
+            alpha_u.value.constant = 0.f;
+            alpha_v = alpha_u;
+        }
+
+        E_NDF ndf;
+        E_SHADOWING_TERM shadowing;
+        E_SCATTER_MODE scatteringMode;
+        SParameter<float> alpha_u;
+        SParameter<float> alpha_v;
+    };
+    struct CMicrofacetDiffuseBSDFNode : CBSDFNode
+    {
+        CMicrofacetDiffuseBSDFNode() : CBSDFNode(ET_MICROFACET_DIFFUSE) {}
+
+        void setSmooth()
+        {
+            alpha_u.source = EPS_CONSTANT;
+            alpha_u.value.constant = 0.f;
+            alpha_v = alpha_u;
+        }
+
+        SParameter<color_t> reflectance;
+        SParameter<float> alpha_u;
+        SParameter<float> alpha_v;
```

Comment:
Oren-Nayar is the only model we have... its isotropic IIRC?

Reply:
but maybe we'll have aniso diffuse some day

Reply:
ok